python/lsst/pipe/base/shims.py

```diff
@@ -0,0 +1,320 @@
+    `~lsst.daf.butler.Butler` backend.
+
+    Parameters
+    ----------
+    butler3 : `lsst.daf.butler.Butler`
+        Generation 3 Butler instance.
+    """
+
+    GENERATION = 2.5
+    """This is a Generation 2 shim for a Generation3 Butler.
+    """
+
+    def __init__(self, butler3):
+        self._butler3 = butler3
+
+    def _makeDataId(self, dataId=None, **rest):
+        """Construct a full data ID by merging the given arguments with the
+        internal data ID.
+        """
+        fullDataId = dict()
+        if dataId is not None:
+            fullDataId.update(dataId)
+        fullDataId.update(rest)
+        return fullDataId
+
+    def _translateDatasetType(self, datasetType):
+        if "_" in datasetType:
+            if datasetType.endswith("_md"):
+                return f"{datasetType[:-3]}.metadata"
+            for component in StorageClassFactory().getStorageClass("Exposure").components:
+                suffix = f"_{component}"
+                if datasetType.endswith(suffix):
+                    return "{}.{}".format(datasetType[:-len(suffix)], component)
+        return datasetType
+
+    def datasetExists(self, datasetType, dataId=None, write=False, **rest):
+        """Check whether a datataset exists in the repository.
+
+        Parameters
+        ----------
+        datasetType : `str`
+            Name of the Gen2 dataset type.
+        dataId : `dict` or `~lsst.daf.butler.DataId`, optional
+            A Generation 3 data ID that identifies the dataset.
+        write : `bool`
+            This option is provided for compatibility with
+            `lsst.daf.persistence.Butler`, but must be `False`.
+        rest
+            Additional key-value pairs to augment the given data ID.
+
+        Returns
+        -------
+        exists : `bool`
+            `True` if the dataset is present in the repository, `False`
+            otherwise.
+        """
+        if write:
+            raise NotImplementedError("ShimButler cannot implement datasetExists with 'write=True'")
+        datasetType = self._translateDatasetType(datasetType)
+        try:
+            return self._butler3.datasetExists(datasetType, self._makeDataId(dataId, **rest))
+        except LookupError:
+            # Gen3 datasetExists raises if Dataset is not present in Registry;
+            # Gen2 does not distinguish between present in Datastore and
+            # present in Registry.
+            return False
+
+    def get(self, datasetType, dataId=None, immediate=True, **rest):
+        """Retrieve a dataset.
+
+        Parameters
+        ----------
+        datasetType : `str`
+            Name of the Gen2 dataset type.
+        dataId : `dict` or `~lsst.daf.butler.DataId`, optional
+            A Generation 3 data ID that identifies the dataset.
+        immediate : `bool`
+            This option is provided for compatibility with
+            `lsst.daf.persistence.Butler`, but must be `True`.
```

Comment:
I suppose in this case that might be better, just because it's possible there is code with `immediate=False` that doesn't really care whether it gets a proxy back.  But I do think we want to check the values of the other not-actually-supported options in this commit, just to make errors earlier and friendlier.